### PR TITLE
feat: allow features to contribute dependencies 

### LINKED ISSUE #655: picocli-gen version is different  from Maven to Gradle
If you select in Micronaut Launch a CLI application with Gradle you get: 

```groovy
annotationProcessor("info.picocli:picocli-codegen:4.2.0")
```

with Maven you get: 

```xml
   <picocli.version>4.6.1</picocli.version>
```

Maven is the correct one.

### LINKED ISSUE #676: Maven project with open api feature adds property micronaut.openapi.version
We generate

```xml
<properties>
    <packaging>jar</packaging>
    <jdk.version>11</jdk.version>
    <release.version>11</release.version>
    <micronaut.version>2.3.4</micronaut.version>
    <exec.mainClass>com.example.Application</exec.mainClass>
    <micronaut.openapi.version>2.3.1</micronaut.openapi.version>
    <micronaut.runtime>netty</micronaut.runtime>
  </properties>
```

but I think we could skip the property `micronaut.openapi.version` and generate just: 

```xml
<properties>
    <packaging>jar</packaging>
    <jdk.version>11</jdk.version>
    <release.version>11</release.version>
    <micronaut.version>2.3.4</micronaut.version>
    <exec.mainClass>com.example.Application</exec.mainClass>
    <micronaut.runtime>netty</micronaut.runtime>
  </properties>
```

The property is set in the bom: 

https://s01.oss.sonatype.org/content/repositories/snapshots/io/micronaut/micronaut-bom/2.3.5-SNAPSHOT/micronaut-bom-2.3.5-20210302.165648-1.pom

### LINKED ISSUE #677: Gradle CLI app old picocli-gen version
Currently a CLI app using gradle uses: 

```groovy
    annotationProcessor("info.picocli:picocli-codegen:4.2.0")
```

it should use: 

```groovy
    annotationProcessor("info.picocli:picocli-codegen:4.6.1")
```

or better just use: 

```groovy
    annotationProcessor("info.picocli:picocli-codegen")
```

and define the version in the bom:

https://github.com/micronaut-projects/micronaut-core/pull/5045

### LINKED ISSUE #678: Dekorate feature for Maven adds unnecessary property
We generate

```xml
<properties>
    <packaging>jar</packaging>
    <jdk.version>11</jdk.version>
    <release.version>11</release.version>
    <micronaut.version>2.3.4</micronaut.version>
    <exec.mainClass>com.example.Application</exec.mainClass>
    <dekorate.version>1.0.3</dekorate.version>
    <micronaut.runtime>netty</micronaut.runtime>
  </properties>
```

but I think we could skip the property `dekorate.version` and generate just: 

```xml
<properties>
    <packaging>jar</packaging>
    <jdk.version>11</jdk.version>
    <release.version>11</release.version>
    <micronaut.version>2.3.4</micronaut.version>
    <exec.mainClass>com.example.Application</exec.mainClass>
    <micronaut.runtime>netty</micronaut.runtime>
  </properties>
```

The property is set in the bom: 

https://s01.oss.sonatype.org/content/repositories/snapshots/io/micronaut/micronaut-bom/2.3.5-SNAPSHOT/micronaut-bom-2.3.5-20210302.165648-1.pom

### PR DESCRIPTION
**Pull request by sdelamo**
This PR allows for:

- Versions to be defined at `starter-core/src/main/resources/pom.xml`  and setup dependabot to keep them up-to-date
- Features contribute dependencies lookups or dependencies
- Features to contribute directly Gradle Plugins

Moreover, it closes:
 
Close: https://github.com/micronaut-projects/micronaut-starter/issues/678
Close: https://github.com/micronaut-projects/micronaut-starter/issues/677
Close: https://github.com/micronaut-projects/micronaut-starter/issues/676
Close: https://github.com/micronaut-projects/micronaut-starter/issues/655

### CODE REVIEW

File: starter-cli/src/test/groovy/io/micronaut/starter/cli/CodeGenConfigSpec.groovy [hunk 1]
```diff
@@ -15,7 +16,9 @@ import java.nio.file.Files
 
 class CodeGenConfigSpec extends Specification {
 
-    @Shared @AutoCleanup BeanContext ctx = BeanContext.run()
+    @Shared
+    @AutoCleanup
+    ApplicationContext ctx = ApplicationContext.run()
```
Comment (sdelamo) [inline starter-cli/src/test/groovy/io/micronaut/starter/cli/CodeGenConfigSpec.groovy:21]: If I leave `BeanContext` fort his test I get: 

```
Failed to inject value for parameter [environment] of class: io.micronaut.http.resource.ResourceLoaderFactory

Message: No bean of type [io.micronaut.context.env.Environment] exists. Make sure the bean is not disabled by bean requirements (enable trace logging for 'io.micronaut.context.condition' to check) and if the bean is enabled then ensure the class is declared a bean and annotation processing is enabled (for Java and Kotlin the 'micronaut-inject-java' dependency should be configured as an annotation processor).
Path Taken: new DefaultAvailableFeatures([List features]) --> new Maven([MavenBuildCreator dependencyResolver]) --> new MavenBuildCreator([MavenCoordinateResolver mavenCoordinateResolver],PropertiesResolver propertiesResolver) --> new MavenCoordinateResolver([DependencyVersionResolver[] dependencyVersionResolvers],PropertiesResolver propertiesResolver) --> new PomDependencyVersionResolver([ResourceResolver resourceResolver]) --> new ResourceLoaderFactory([Environment environment])
io.micronaut.context.exceptions.DependencyInjectionException: Failed to inject value for parameter [environment] of class: io.micronaut.http.resource.ResourceLoaderFactory

Message: No bean of type [io.micronaut.context.env.Environment] exists. Make sure the bean is not disabled by bean requirements (enable trace logging for 'io.micronaut.context.condition' to check) and if the bean is enabled then ensure the class is declared a bean and annotation processing is enabled (for Java and Kotlin the 'micronaut-inject-java' dependency should be configured as an annotation processor).
Path Taken: new DefaultAvailableFeatures([List features]) --> new Maven([MavenBuildCreator dependencyResolver]) --> new MavenBuildCreator([MavenCoordinateResolver mavenCoordinateResolver],PropertiesResolver propertiesResolver) --> new MavenCoordinateResolver([DependencyVersionResolver[] dependencyVersionResolvers],PropertiesResolver propertiesResolver) --> new PomDependencyVersionResolver([ResourceResolver resourceResolver]) --> new ResourceLoaderFactory([Environment environment])
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1035)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1924)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2705)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2691)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2363)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2337)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1206)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1924)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2705)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2691)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2363)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2337)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1245)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1013)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1924)
	at io.micronaut.context.DefaultBeanContext.addCandidateToList(DefaultBeanContext.java:3048)
	at io.micronaut.context.DefaultBeanContext.getBeansOfTypeInternal(DefaultBeanContext.java:2949)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1045)
	at io.micronaut.context.AbstractBeanDefinition.lambda$getBeansOfTypeForConstructorArgument$9(AbstractBeanDefinition.java:1143)
	at io.micronaut.context.AbstractBeanDefinition.resolveBeanWithGenericsFromConstructorArgument(AbstractBeanDefinition.java:1822)
	at io.micronaut.context.AbstractBeanDefinition.getBeansOfTypeForConstructorArgument(AbstractBeanDefinition.java:1138)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:987)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1924)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2705)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2691)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2363)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2337)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1245)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1013)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1924)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2705)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2691)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2363)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2337)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1245)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1013)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1924)
	at io.micronaut.context.DefaultBeanContext.addCandidateToList(DefaultBeanContext.java:3048)
	at io.micronaut.context.DefaultBeanContext.getBeansOfTypeInternal(DefaultBeanContext.java:2949)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1045)
	at io.micronaut.context.AbstractBeanDefinition.lambda$getBeansOfTypeForConstructorArgument$9(AbstractBeanDefinition.java:1143)
	at io.micronaut.context.AbstractBeanDefinition.resolveBeanWithGenericsFromConstructorArgument(AbstractBeanDefinition.java:1822)
	at io.micronaut.context.AbstractBeanDefinition.getBeansOfTypeForConstructorArgument(AbstractBeanDefinition.java:1138)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:990)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1924)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2705)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2691)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2363)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2337)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:721)
	at io.micronaut.starter.cli.CodeGenConfig.load(CodeGenConfig.java:155)
	at io.micronaut.starter.cli.CodeGenConfigSpec.test config is compatible with the old format(CodeGenConfigSpec.groovy:35)
Caused by: io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [io.micronaut.context.env.Environment] exists. Make sure the bean is not disabled by bean requirements (enable trace logging for 'io.micronaut.context.condition' to check) and if the bean is enabled then ensure the class is declared a bean and annotation processing is enabled (for Java and Kotlin the 'micronaut-inject-java' dependency should be configured as an annotation processor).
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2348)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1245)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1013)
	... 51 more
```

Comment (jameskleeh) [reply starter-cli/src/test/groovy/io/micronaut/starter/cli/CodeGenConfigSpec.groovy:21]: Yep to be expected. The environment, etc.. is not part of the core bean context

**Review by sdelamo — COMMENTED** [review]
@jameskleeh it would be helpful to get another set of 👀

**Comment by sdelamo** [comment]
Closed via https://github.com/micronaut-projects/micronaut-starter/pull/683

**Pull request closed without merging**